crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -11,48 +11,133 @@
 
 use super::{definition_expression_type, DynamicType, Type};
+use crate::semantic_index::definition::Definition;
+use crate::symbol::Boundness;
+use crate::types::todo_type;
 use crate::Db;
-use crate::{semantic_index::definition::Definition, types::todo_type};
 use ruff_python_ast::{self as ast, name::Name};
 
+/// The signature of a possible union of callables.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, salsa::Update)]
+pub(crate) struct Signatures<'db> {
+    /// The type that is (hopefully) callable.
+    pub(crate) ty: Type<'db>,
+    /// For an object that's callable via a `__call__` method, the type of that method. For an
+    /// object that is directly callable, the type of the object.
+    pub(crate) signature_ty: Type<'db>,
+    elements: Vec<CallableSignature<'db>>,
+}
+
+impl<'db> Signatures<'db> {
+    pub(crate) fn not_callable(ty: Type<'db>, signature_ty: Type<'db>) -> Self {
+        Self {
+            ty,
+            signature_ty,
+            elements: vec![CallableSignature::not_callable(ty, signature_ty)],
+        }
+    }
+
+    pub(crate) fn single(signature: CallableSignature<'db>) -> Self {
+        Self {
+            ty: signature.ty,
+            signature_ty: signature.signature_ty,
+            elements: vec![signature],
+        }
+    }
+
+    /// Creates a new `Signatures` from an iterator of [`Signature`]s. Panics if the iterator is
+    /// empty.
+    pub(crate) fn from_union<I>(ty: Type<'db>, signature_ty: Type<'db>, elements: I) -> Self
+    where
+        I: IntoIterator<Item = &'db Signatures<'db>>,
+    {
+        let elements = elements
+            .into_iter()
+            .flat_map(|s| s.elements.iter().cloned())
+            .collect();
+        Self {
+            ty,
+            signature_ty,
+            elements,
+        }
+    }
+
+    pub(crate) fn as_single(&self) -> Option<&CallableSignature<'db>> {
+        match self.elements.as_slice() {
+            [signature] => Some(signature),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = &CallableSignature<'db>> {
+        self.elements.iter()
+    }
+
+    pub(crate) fn set_dunder_call_boundness(&mut self, boundness: Boundness) {
+        for signature in &mut self.elements {
+            signature.dunder_call_boundness = Some(boundness);
+        }
+    }
+}
+
 /// The signature of a single callable. If the callable is overloaded, there is a separate
 /// [`Signature`] for each overload.
 #[derive(Clone, Debug, PartialEq, Eq, Hash, salsa::Update)]
-pub enum CallableSignature<'db> {
-    Single(Signature<'db>),
-    Overloaded(Box<[Signature<'db>]>),
+pub(crate) struct CallableSignature<'db> {
+    /// The type that is (hopefully) callable.
+    pub(crate) ty: Type<'db>,
+
+    /// For an object that's callable via a `__call__` method, the type of that method. For an
+    /// object that is directly callable, the type of the object.
+    pub(crate) signature_ty: Type<'db>,
```

Comment:
nit: we changed our naming convention in https://github.com/astral-sh/ruff/pull/15617 to use `type` rather than `ty` where possible. That obviously isn't ergonomic for the first field in this struct (as it conflicts with a Rust keyword), but I'd prefer for this to be called `signature_type` (and similar for various other fields on various other structs in this PR)

```suggestion
    pub(crate) signature_type: Type<'db>,
```

Reply:
Done (and `ty` → `callable_type`)